base_delivery_carrier_files/carrier_file.py

```diff
@@ -0,0 +1,153 @@
+#    it under the terms of the GNU Affero General Public License as
+#    published by the Free Software Foundation, either version 3 of the
+#    License, or (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+import os
+import logging
+
+from openerp import models, fields, api, exceptions
+from openerp.tools.translate import _
+from .generator import new_file_generator
+
+
+class CarrierFile(models.Model):
+    _name = 'delivery.carrier.file'
+
+    @api.model
+    def get_type_selection(self):
+        """
+        Has to be inherited to add carriers
+        """
+        return [('generic', 'Generic')]
+
+    @api.model
+    def get_write_mode_selection(self):
+        """
+        Selection can be inherited to add more write modes
+        """
+        return [('disk', 'Disk')]
+
+    @api.multi
+    def _write_file(self, filename, file_content):
+        """
+        Method responsible of writing the file, on the filesystem or
+        by inheriting the module, in the document module as instance
+
+        :param browse_record carrier_file: browsable carrier.file
+                                           (configuration)
+        :param tuple filename: name of the file to write
+        :param tuple file_content: content of the file to write
+        :return: True if write is successful
+        """
+        for carrier_file in self:
+            if not carrier_file.export_path:
+                raise exceptions.except_orm(_('Error'),
+                                            _('Export path is not defined '
+                                            'for carrier file %s') %
+                                            (carrier_file.name,))
+            full_path = os.path.join(carrier_file.export_path, filename)
+            with open(full_path, 'w') as file_handle:
+                file_handle.write(file_content)
+        return True
+
+    @api.one
+    def _generate_files(self, picking_ids):
+        """
+        Generate one or more files according to carrier_file configuration
+        for all picking_ids
+
+        :param browse_record carrier_file: browsable carrier file
+                                           configuration
+        :param list picking_ids: list of ids of pickings for which
+                                 we have to generate a file
+        :return: True if successful
+        """
+        # Convert frozendict context to editable context
+        context = dict(self.env.context or {})
```

Comment:
Why do you copy entire self.env.context if it is to only change key `picking_id` afterward?

I think you could only define an empty dict here.